### PR TITLE
feat(hook): add optimizeModules hook

### LINKED ISSUE #2423: [Feature Request]: support compilation.hooks.optimizeModules
### What problem does this feature solve?

We need `compilation.hooks.optimizeModules` to be compatible with [circular-dependency-plugin](https://github.com/aackerman/circular-dependency-plugin)

### What does the proposed API of configuration look like?

``` js
compilation.hooks.optimizeModules.tap("PluginName", (modules) => {})
```

### PR DESCRIPTION
**Pull request by nieyuyao**
## Related issue (if exists)

fix #2423

### CODE REVIEW

File: packages/rspack/src/compiler.ts [hunk 1]
```diff
@@ -364,6 +365,9 @@ class Compiler {
 		);
 		this.#updateDisabledHooks();
 	}
+	async #optimize_modules() {
+		await this.compilation.hooks.optimizeModules.promise(this.compilation.getModules())
```
Comment (IWANABETHATGUY) [inline packages/rspack/src/compiler.ts:369]: pelase add `this.#updateDisabledHooks();` after this line.

**Comment by IWANABETHATGUY** [comment]
looks good to me, @jerrykingxyz  would you mind having a look if this pr fits your needs?

File: packages/rspack/src/compiler.ts (hunk 1 above)
Comment (jerrykingxyz) [reply packages/rspack/src/compiler.ts:369]: and this hook needs to be added to the hookMap in `#updateDisabledHooks`

Comment (IWANABETHATGUY) [reply packages/rspack/src/compiler.ts:369]: @nieyuyao do you have time to address this review issue? ask me anything if you meet any problems.

Comment (nieyuyao) [reply packages/rspack/src/compiler.ts:369]: Sorry,  I've been a bit busy lately, I will address the issue as soon as possible.

**Comment by IWANABETHATGUY** [comment]
One more thing,  you should add your newly added hooks here to make the hooks filter work, https://github.com/web-infra-dev/rspack/pull/2758/files#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdR330-R331

**Review by IWANABETHATGUY — APPROVED** [review]
LGTM, thanks!

**Pull request merged**